### DISCUSSION TITLE
xmidt clustering

### DISCUSSION THREAD

**Discussion by outbackdingo** [xmidt-org/xmidt#76] [General]
quick question, kubernetes xmit clustering for 5M devices, in everyones opinion whats the best front end for scaling and has anyone benchmarked say single node scalability ? trying to overcome the C10M problem without 300 front end servers.... enlighten us please. Thanks in advance.

**Comment by schmidtw** [comment]
@Sachin4403 do you have any suggestions?

My team doesn't run in kubernetes, but is using EC2/VMs to avoid this group of problems.

**Reply by outbackdingo** [reply]
not sure how your going to scale to 5M devices in AWS without some magic also, please enlighten me

**Comment by Sachin4403** [comment]
Hello @outbackdingo 

Since the device will be connected to Talaria and so we must have only one Talaria per service and deployment so that Scytale can find the same and fanout the traffic to that Talaria, On the other hand, Kubernetes Pod capacity will be completely dependent on the CPU and Memory which is configured in Kubernetes. rest all services will be scaled  under a single Kubernetes services

On the other hand, Kubernetes internally uses the service name to re-route the traffic but when you give the same response to the device then the device will not be able to connect back because the device is not aware of that service name in its network. for this, we are using a [petasos-rewriter](https://github.com/Sachin4403/petasos-rewriter) in front of petasos so that all of our traffic will be rerouted via private services.

Do let me know if you need a more detailed understanding of it.

**Reply by outbackdingo** [reply]
my curiousity is about device count, so are we loadbalancing or other in front of taleria to enable it to scale to such numbers? I cant see 1 taleria process handling 1M 5M obviously taleria needs to be able to scale horizontally, so what piece of the architecture allows for say X talerias running to accommodate 5M devices ?? metallb or haproxy type configuration comes to my mind, it that a viable working model to scale a k8s environment. Has there ever been a scalability test? what was the max clients compared to say sizing of the cluster or even single machine. How can we "size" the cluster needed and resources?
